posts/2022-06-27-foldleft-introduction-en.adoc

```diff
@@ -0,0 +1,372 @@
+          }
+        }
+        .leftMap(error => new IllegalArgumentException(s"Unable to delete the task due to ${error.getMessage}"))
+        .toEither
+    }
+  } yield updatedPlayers
+Await.result(resultEitherT.value, 1.second)
+/* Result =>
+Left(java.lang.IllegalArgumentException: Unable to delete the task due to status is Enrolled)
+*/
+----
+
+Précisions :
+
+[source, scala]
+----
+acc.andThen { players =>
+            current.map(_ +: players)
+          }
+----
+
+Again, this is the Cats version of left.map(...)
+Finally, we wrap our block with `EitherT.fromEither[Future] { { ... }.toEither }` to change the type from `Validated` to `EitherT`.
+
+== Performances
+
+If you remember, I proposed this implementation at the beginning of the article :
+
+[source, scala]
+----
+val numbers = List(1, 2, 3, 4)
+val newList: Seq[Int] = numbers.foldLeft[Seq[Int]](Seq.empty) {
+    case (accumulator, number) if accumulator.isEmpty => accumulator :+ number
+    case (accumulator, number) => accumulator :+ (accumulator.last + number)
+}
+println(newList)
+// Result => List(1, 3, 6, 10)
+----
+
+In reality, we have initialized the accumulator with `Seq.empty[Int]` while the proposed list is of type `List`. Since the type is generic (foldLeft takes a `Seq[Int]`), the compiler will assign the `List` type of our list of numbers to the accumulator. The problem for performance is that the `List` type will be found at every stage of processing the list elements:
+
+1. when retrieving the last element of the list (`accumulator.last`)
+
+2. when adding the new element to the end of the list (`accumulator :+ ...`)
+
+For the `accumulator.last`, the implementation in Scala deletes the first element, then looks at how many elements are left in the list. It will do this again until there is only one element left to return.
+
+Referring to the scala documentation on https://docs.scala-lang.org/overviews/collections-2.13/performance-characteristics.html[collection performance], the operation of adding items to a `List` collection takes longer the larger the list.
+
+To solve this problem while remaining generic, we can rewrite our method as follows :
+
+[source, scala]
+----
+List(1,2,3,4).foldLeft(Seq.empty[Int]) {
+  case (Nil, element) => Seq(element)
+  case (accumulator, element) => (accumulator.head + element) +: accumulator
+}.reverse
+----
+
+In this way, accessing or adding an element to the accumulator will be a constant time (fast) operation. But as this implementation produces a reverse result, it is sufficient to add a `.reverse` at the end, or to use a foldRight, which you will notice is just a reverse foldLeft.
+
+`def foldRight[B](z: B)(op: (A, B) => B): B = reversed.foldLeft(z)((b, a) => op(a, b))`
+
+== Conclusion
+
+As we have seen, *foldLeft* and *foldRight* are kind of all-purpose boxes when you want to go through a list in a given order and compare the entries with each other while returning a different type if needed. 
```

Comment:
Maybe you can rephrase this as:

As demonstrated in this article, *foldLeft* and *foldRight* are very powerful methods that can be considered the equivalent of the Swiss Army knife in the Scala collection library: it operates on a collection of elements of some type A and can generate a value that is the same type A or a completely different type B.